File: Problem_for_the_Day/PathSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void helper(TreeNode* root, int targetSum, vector<vector<int>> &res, vector<int> &arr){
        if(!root){
            return;
        }
        arr.push_back(root->val);
        targetSum-=root->val;
        if(!root->left && !root->right){
            //check and add to final res when we reach leaf
            if(targetSum==0){
                res.push_back(arr);
            }
        }
            else{
                helper(root->left,targetSum,res,arr);
                helper(root->right,targetSum,res,arr);
            }
        //if targetSum != 0 then remove 
           arr.pop_back();
    }
    
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        //dfs
        vector<int> arr;
        vector<vector<int>> res;
    
        helper(root,targetSum,res,arr);
        return res;
    }
};

File: Tree/PathSum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if root == None:
            return False
        elif root.val ==targetSum and root.left == None and root.right == None:
            return True
        else:
            target = targetSum - root.val
            return self.hasPathSum(root.left, target) or self.hasPathSum(root.right, target) 
